src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class ReadOnlyTagHelperAttributes<TAttributeType> : IReadOnlyList<TAttributeType>
+        where TAttributeType : IReadOnlyTagHelperAttribute
+    {
+        protected readonly List<TAttributeType> _attributes;
+
+        protected ReadOnlyTagHelperAttributes()
+        {
+            _attributes = new List<TAttributeType>();
+        }
+
+        public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttributeType> attributes)
+        {
+            _attributes = new List<TAttributeType>(attributes);
+        }
+
+        public TAttributeType this[int index]
+        {
+            get
+            {
+                return _attributes[index];
+            }
+        }
+
+        public TAttributeType this[[NotNull] string key]
```

Comment:
Ehh, what's the value in removing this though? It makes accessing/setting elements for the 90% case (no multiple keys same name) significantly easier. The problem with the `IDictionary` approach that @Eilon was complaining about (I believe) was more geared towards adding methods like `Add(string key, TAttribute attribute)` which don't really make sense in this context.

I'd be fine with removing the setter (which then enables us to remove the implicit string => TAttribute converter). That would remove the dictionary like concept of adding an entry with a key instead of the entry itself being essentially a better kvp.